File: oxidize-golang/internal/generate/generate.go
package generate

import "github.com/Zapdev-labs/oxidize/golang/internal/api"

type PlaceholderSpec struct {
	ResponseFormat *api.ResponseFormat
	GuidedJSON     []byte
	JSONSchema     []byte
	GuidedRegex    string
	GuidedChoice   []string
}

func PlaceholderText(spec PlaceholderSpec) string {
	if len(spec.GuidedChoice) > 0 {
		return spec.GuidedChoice[0]
	}
	if len(spec.GuidedJSON) > 0 || len(spec.JSONSchema) > 0 {
		return "{}"
	}
	if spec.GuidedRegex != "" {
		return spec.GuidedRegex
	}
	if spec.ResponseFormat != nil {
		return spec.ResponseFormat.OutputText()
	}
	return ""
}

func CLIText(prompt string) string {
	return "oxidize-cli: " + prompt
}

func CLITranscript(prompt string) string {
	return "generation progress: 1/2 tokens\ngeneration progress: 2/2 tokens\n" + CLIText(prompt) + "\ngeneration stats: tokens=2 speed=4.00 tok/s\n"
}

File: oxidize-python/oxidize_python/internal/generate.py
from dataclasses import dataclass


@dataclass
class PlaceholderSpec:
    response_format: object | None = None
    guided_json: bytes | None = None
    json_schema: bytes | None = None
    guided_regex: str = ""
    guided_choice: list[str] | None = None


def placeholder_text(spec: PlaceholderSpec) -> str:
    if spec.guided_choice:
        return spec.guided_choice[0]
    if spec.guided_json or spec.json_schema:
        return "{}"
    if spec.guided_regex:
        return spec.guided_regex
    if spec.response_format is not None and hasattr(spec.response_format, "output_text"):
        return spec.response_format.output_text()
    return ""


def cli_text(prompt: str) -> str:
    return f"oxidize-cli: {prompt}"


def cli_transcript(prompt: str) -> str:
    return (
        "generation progress: 1/2 tokens\n"
        "generation progress: 2/2 tokens\n"
        f"{cli_text(prompt)}\n"
        "generation stats: tokens=2 speed=4.00 tok/s\n"
    )
